File: Tests/SelfTest/StorageTests.swift
import ClipDeckKit
import Foundation

func runStorageTests() {
    func makeTempDirectory() -> URL {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("clipdeck-test-\(UUID().uuidString)")
        try? FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        return url
    }

    Harness.suite("LibraryStorage — primeira execução") {
        let dir = makeTempDirectory()
        defer { try? FileManager.default.removeItem(at: dir) }
        let storage = LibraryStorage(directory: dir)

        guard let result = try? storage.load() else {
            return Harness.expect(false, "load() não deveria lançar")
        }
        Harness.expectEqual(result.outcome, .seededNewFile, "semeia quando não há arquivo")
        Harness.expect(!result.library.folders.isEmpty, "vem com pastas de exemplo")
        Harness.expect(
            FileManager.default.fileExists(atPath: storage.fileURL.path),
            "grava o arquivo no primeiro load"
        )
    }

    Harness.suite("LibraryStorage — roundtrip") {
        let dir = makeTempDirectory()
        defer { try? FileManager.default.removeItem(at: dir) }
        let storage = LibraryStorage(directory: dir)

        let folder = Folder(name: "Cobrança", order: 3)
        let template = Template(
            title: "Aviso de vencimento",
            body: "Oi {{nome}}, sua fatura vence em {{data}}.",
            folderID: folder.id,
            tags: ["urgente", "financeiro"],
            useCount: 7
        )
        let original = Library(folders: [folder], templates: [template])

        try? storage.save(original)
        guard let reloaded = try? storage.load().library else {
            return Harness.expect(false, "load() após save não deveria lançar")
        }

        Harness.expectEqual(reloaded.templates.count, 1, "preserva a contagem de templates")
        Harness.expectEqual(reloaded.templates[0].title, "Aviso de vencimento", "preserva o título")
        Harness.expectEqual(reloaded.templates[0].useCount, 7, "preserva o contador de uso")
        Harness.expectEqual(reloaded.templates[0].tags, ["urgente", "financeiro"], "preserva as tags")
        Harness.expectEqual(reloaded.templates[0].folderID, folder.id, "preserva o vínculo com a pasta")
        Harness.expectEqual(reloaded.folders[0].name, "Cobrança", "preserva acentuação no nome da pasta")
    }

    Harness.suite("LibraryStorage — recuperação de arquivo corrompido") {
        let dir = makeTempDirectory()
        defer { try? FileManager.default.removeItem(at: dir) }
        let storage = LibraryStorage(directory: dir)

        // Grava duas vezes: a segunda gera o backup da primeira.
        let good = Library(templates: [Template(title: "Bom template", body: "conteúdo")])
        try? storage.save(good)
        try? storage.save(good)

        // Simula corrupção (queda de energia, edição manual errada, sync ruim).
        try? Data("{ isso não é json".utf8).write(to: storage.fileURL)

        guard let result = try? storage.load() else {
            return Harness.expect(false, "load() não deveria lançar com arquivo corrompido")
        }

        if case .recoveredFromBackup = result.outcome {
            Harness.expect(true, "recupera a partir do backup mais recente")
        } else {
            Harness.expect(false, "deveria recuperar do backup, obteve \(result.outcome)")
        }
        Harness.expectEqual(result.library.templates.first?.title, "Bom template", "conteúdo volta intacto")

        let quarantined = (try? FileManager.default.contentsOfDirectory(atPath: dir.path))?
            .filter { $0.hasPrefix("library.corrupt-") } ?? []
        Harness.expect(!quarantined.isEmpty, "preserva o arquivo corrompido em quarentena")
    }

    Harness.suite("LibraryStorage — corrompido sem backup") {
        let dir = makeTempDirectory()
        defer { try? FileManager.default.removeItem(at: dir) }
        let storage = LibraryStorage(directory: dir)

        try? Data("lixo".utf8).write(to: storage.fileURL)
        guard let result = try? storage.load() else {
            return Harness.expect(false, "load() não deveria lançar")
        }

        if case .seededAfterCorruption = result.outcome {
            Harness.expect(true, "semeia de novo quando não há backup aproveitável")
        } else {
            Harness.expect(false, "esperava seededAfterCorruption, obteve \(result.outcome)")
        }
    }

    Harness.suite("LibraryStorage — rotação de backups") {
        let dir = makeTempDirectory()
        defer { try? FileManager.default.removeItem(at: dir) }
        let storage = LibraryStorage(directory: dir)

        for index in 0..<9 {
            try? storage.save(Library(templates: [Template(title: "v\(index)", body: "x")]))
        }

        let backups = (try? FileManager.default.contentsOfDirectory(atPath: storage.backupsDirectory.path))?
            .filter { $0.hasPrefix("library-") } ?? []
        Harness.expect(backups.count <= 5, "mantém no máximo 5 backups (obteve \(backups.count))")
        Harness.expect(backups.count >= 1, "mantém pelo menos 1 backup")
    }

    Harness.suite("Decodificação tolerante") {
        let partial = """
        { "templates": [ { "title": "Só título" } ] }
        """
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601

        let library = try? decoder.decode(Library.self, from: Data(partial.utf8))
        Harness.expect(library != nil, "carrega JSON escrito à mão sem campos opcionais")
        Harness.expectEqual(library?.templates.first?.title, "Só título", "usa o título informado")
        Harness.expectEqual(library?.templates.first?.useCount, 0, "usa 0 como padrão de useCount")
        Harness.expectEqual(library?.folders.count, 0, "aceita ausência de pastas")
    }
}

File: windows/tests/ClipDeck.Core.Tests/StorageTests.cs
using System.Text.Json;
using ClipDeck.Core.Models;

namespace ClipDeck.Core.Tests;

public class LibraryStorageTests : IDisposable
{
    private readonly string _dir;
    private readonly LibraryStorage _storage;

    public LibraryStorageTests()
    {
        _dir = Path.Combine(Path.GetTempPath(), "clipdeck-test-" + Guid.NewGuid());
        _storage = new LibraryStorage(_dir);
    }

    public void Dispose()
    {
        try { Directory.Delete(_dir, recursive: true); } catch { /* melhor esforço */ }
    }

    [Fact]
    public void SemeiaQuandoNaoHaArquivo()
    {
        var result = _storage.Load();
        Assert.Equal(LoadOutcomeKind.SeededNewFile, result.Outcome.Kind);
        Assert.NotEmpty(result.Library.Folders);
        Assert.True(File.Exists(_storage.FilePath));
    }

    [Fact]
    public void RoundtripPreservaTudo()
    {
        var folder = new Folder { Name = "Cobrança", Order = 3 };
        var template = new Template
        {
            Title = "Aviso de vencimento",
            Body = "Oi {{nome}}, sua fatura vence em {{data}}.",
            FolderId = folder.Id,
            Tags = { "urgente", "financeiro" },
            UseCount = 7,
        };

        _storage.Save(new Library { Folders = { folder }, Templates = { template } });
        var reloaded = _storage.Load().Library;

        var loaded = Assert.Single(reloaded.Templates);
        Assert.Equal("Aviso de vencimento", loaded.Title);
        Assert.Equal(7, loaded.UseCount);
        Assert.Equal(new[] { "urgente", "financeiro" }, loaded.Tags);
        Assert.Equal(folder.Id, loaded.FolderId);
        Assert.Equal("Cobrança", reloaded.Folders[0].Name);
    }

    [Fact]
    public void RecuperaDeBackupQuandoArquivoCorrompe()
    {
        var good = new Library { Templates = { new Template { Title = "Bom template", Body = "conteúdo" } } };
        _storage.Save(good);
        _storage.Save(good); // a segunda gravação gera o backup da primeira

        File.WriteAllText(_storage.FilePath, "{ isso não é json");

        var result = _storage.Load();
        Assert.Equal(LoadOutcomeKind.RecoveredFromBackup, result.Outcome.Kind);
        Assert.Equal("Bom template", result.Library.Templates[0].Title);
        Assert.NotEmpty(Directory.GetFiles(_dir, "library.corrupt-*.json"));
    }

    [Fact]
    public void SemeiaQuandoCorrompidoESemBackup()
    {
        Directory.CreateDirectory(_dir);
        File.WriteAllText(_storage.FilePath, "lixo");
        Assert.Equal(LoadOutcomeKind.SeededAfterCorruption, _storage.Load().Outcome.Kind);
    }

    [Fact]
    public void MantemNoMaximoCincoBackups()
    {
        for (var i = 0; i < 9; i++)
        {
            _storage.Save(new Library { Templates = { new Template { Title = $"v{i}", Body = "x" } } });
        }

        var backups = Directory.GetFiles(_storage.BackupsDirectory, "library-*.json");
        Assert.InRange(backups.Length, 1, 5);
    }

    [Fact]
    public void CarregaJsonEscritoAMaoSemCamposOpcionais()
    {
        var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
        var library = JsonSerializer.Deserialize<Library>(
            """{ "templates": [ { "title": "Só título" } ] }""", options);

        Assert.NotNull(library);
        Assert.Equal("Só título", library!.Templates[0].Title);
        Assert.Equal(0, library.Templates[0].UseCount);
        Assert.Empty(library.Folders);
    }
}
